registrations/api.py

```diff
@@ -254,6 +260,90 @@ def none_to_unlim(val):
 
             return Response(data, status=status.HTTP_201_CREATED)
 
+    @action(
+        methods=["post"],
+        detail=True,
+        permission_classes=[DataSourceResourceEditPermission],
+    )
+    def send_message(self, request, pk=None, version=None):
+        registration = self.get_object()
+
+        errors = {}
+
+        # Validate that required fields are filled
+        required_fields = ["subject", "body"]
+        for field in required_fields:
+            if not request.data.get(field):
+                errors[field] = _("This field must be specified.")
+
+        # Send message to defined signups if signups attribute is defined and it's an array
+        # By default send message to all signups
+        if request.data.get("signups") and hasattr(
+            request.data.get("signups"), "__len__"
+        ):
+            try:
+                signups = registration.signups.filter(
+                    pk__in=[i for i in request.data.get("signups")]
+                )
+            except ValueError as error:
+                errors["signups"] = str(error)
+
+        else:
+            signups = registration.signups.all()
+        signups.exclude(email=None)
+
+        # Return validations errors if there is any
+        if errors:
+            raise serializers.ValidationError(errors)
```

Comment:
Data validation should be something that a serializer does, not a view.

Reply:
More of an idea rather than working code, but instead of doing what a serializer does you could do something like:

```
# Add a serializer class for the email
class MassEmailSerializer(serializers.Serializer):
    subject = serializers.CharField(required=True)
    body = serializers.CharField(required=True)
    signups = serializers.ListField(child=serializers.CharField(), allow_empty=False, min_length=1,)


# Add this to the viewset to return a correct serializer for the send_message action 
def get_serializer_class(
    if self.action == 'send_message':
        return MassEmailSerializer
    return super().get_serializer_class()


# in this action method you'd then do something like

serializer = self.get_serializer(data=request.data)
serializer.is_valid(raise_exception=True)

# Get values from serializer.validated_data
body = serializer.validated_data["body"]
```

Reply:
Thanks guys. I added MassEmailSerializer serializer to validate send_message action payload as suggested